Mathlib/LinearAlgebra/TensorProduct/Submodule.lean

```diff
@@ -0,0 +1,202 @@
+  simp_rw [LinearMap.comp_apply, LinearEquiv.coe_coe, TensorProduct.comm_tmul, mulMap_tmul]
+  exact (hc m n).symm
+
+theorem mulMap_comp_rTensor (M' : Submodule R S) (hM : M' ≤ M) :
+    mulMap M N ∘ₗ (inclusion hM).rTensor N = mulMap M' N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_lTensor (N' : Submodule R S) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ (inclusion hN).lTensor M = mulMap M N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_map_inclusion (M' N' : Submodule R S) (hM : M' ≤ M) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ TensorProduct.map (inclusion hM) (inclusion hN) = mulMap M' N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_eq_mul'_comp_mapIncl : mulMap M N = .mul' R S ∘ₗ TensorProduct.mapIncl M N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_range : LinearMap.range (mulMap M N) = M * N := by
+  refine le_antisymm ?_ (mul_le.2 fun m hm n hn ↦ ⟨⟨m, hm⟩ ⊗ₜ[R] ⟨n, hn⟩, rfl⟩)
+  rintro _ ⟨x, rfl⟩
+  induction x using TensorProduct.induction_on with
+  | zero => rw [_root_.map_zero]; exact zero_mem _
+  | tmul a b => exact mul_mem_mul a.2 b.2
+  | add a b ha hb => rw [_root_.map_add]; exact add_mem ha hb
+
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] M * N` induced by multiplication in `S`,
+which is surjective (`Submodule.mulMap'_surjective`). -/
+def mulMap' : M ⊗[R] N →ₗ[R] ↥(M * N) :=
+  (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
+
+@[simp]
+theorem val_mulMap'_tmul (m : M) (n : N) : (mulMap' M N (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
+
+theorem mulMap'_surjective : Function.Surjective (mulMap' M N) := by
+  simp_rw [mulMap', LinearMap.coe_comp, LinearEquiv.coe_coe, EquivLike.comp_surjective,
+    LinearMap.surjective_rangeRestrict]
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural map
+`i(R) ⊗[R] N →ₗ[R] N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This is promoted to an isomorphism as `Submodule.lTensorOne`. Use that instead. -/
+def lTensorOne' : (1 : Submodule R S) ⊗[R] N →ₗ[R] N :=
+  (LinearEquiv.ofEq _ _ (by rw [mulMap_range, one_mul])).toLinearMap ∘ₗ (mulMap _ N).rangeRestrict
+
+theorem lTensorOne'_apply
+    (y : R) {hy : algebraMap R S y ∈ (1 : Submodule R S)}
+    (n : N) : N.lTensorOne' (⟨algebraMap R S y, hy⟩ ⊗ₜ[R] n) = y • n :=
+  Subtype.val_injective <| by simp [lTensorOne', Algebra.smul_def]
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural isomorphism between
+`i(R) ⊗[R] N` and `N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This generalizes `TensorProduct.lid` as `i(R)` is not necessarily isomorphic to `R`. -/
+def lTensorOne : (1 : Submodule R S) ⊗[R] N ≃ₗ[R] N := by
+  refine LinearEquiv.ofLinear N.lTensorOne' (TensorProduct.mk R (1 : Submodule R S) N
+    ⟨1, one_le.1 (le_refl _)⟩) (by ext; simp [lTensorOne']) (TensorProduct.ext' fun r n ↦ ?_)
```

Comment:
I'm sorry, but my opinion is keep using `(1 : Submodule R S)` since this file is for submodule, not subalgebra. I'll mark this conversation as resolved. Feel free to unresolve it if there are further comments.

Reply:
Sorry forgot about this PR, but my main concern is that the simp normal form should not be `⟨algebraMap R S y, algebraMap_mem y⟩` and `⟨1, one_le.1 (le_refl _)⟩`. A way to avoid this is to define `One (1 : Submodule R S)` and custom `algebraMapToOne`, but they will need some lemmas to set up, and I truly believe the easier way is to reuse the subalgbra API here.

Reply:
OK I see your concern. Let me see if it's better to rename it to `(l|r)TensorBot`.

Reply:
Done.